bpfd/src/bpf.rs

```diff
@@ -41,18 +41,145 @@ const MAPS_MODE: u32 = 0o0660;
 
 pub(crate) struct BpfManager {
     config: Config,
-    dispatchers: HashMap<DispatcherId, Dispatcher>,
-    programs: HashMap<Uuid, Program>,
+    dispatchers: BpfDispatchers,
+    programs: BpfPrograms,
     maps: HashMap<Uuid, BpfMap>,
     commands: mpsc::Receiver<Command>,
 }
 
+pub(crate) struct BpfPrograms {
+    programs: HashMap<Uuid, Program>,
+}
+
+impl BpfPrograms {
+    fn insert(&mut self, id: Uuid, prog: &Program) -> Option<Program> {
+        let prog = prog.clone();
+        self.programs.insert(id, prog)
+    }
+
+    fn remove(&mut self, id: &Uuid) -> Option<Program> {
+        self.programs.remove(id)
+    }
+
+    fn get_mut(&mut self, id: &Uuid) -> Option<&mut Program> {
+        self.programs.get_mut(id)
+    }
+
+    fn get(&self, id: &Uuid) -> Option<&Program> {
+        self.programs.get(id)
+    }
+
+    fn contains_key(&self, id: &Uuid) -> bool {
+        self.programs.contains_key(id)
+    }
+
+    fn next_avail_id(&self, program: &Program) -> usize {
+        self.programs
+            .iter()
+            .filter(|(_, p)| {
+                if p.kind() == program.kind() {
+                    p.if_index() == program.if_index() && p.direction() == program.direction()
+                } else {
+                    false
+                }
+            })
+            .collect::<HashMap<_, _>>()
+            .len()
+    }
+
+    fn collect_programs(
+        &mut self,
+        program_type: ProgramType,
+        if_index: Option<u32>,
+        direction: Option<Direction>,
+    ) -> Vec<(&Uuid, &mut Program)> {
```

Comment:
In a local branch I played around with the return type being `impl Iterator<Item = (&Uuid, &mut Program)>` which lets us avoid allocating (via `collect()) here.

Reply:
Let me give this a try.

Reply:
Done just needed to add some custom lifetime specifiers.